harness/determined/cli/command.py

```diff
@@ -282,15 +284,53 @@ def config(args: Namespace) -> None:
     print(render.format_object_as_yaml(res_json["config"]))
 
 
-def _set_nested_config(config: Dict[str, Any], key_path: List[str], value: Any) -> Dict[str, Any]:
+# Convert config overrides in dot notation to json path
+# For example the caller of this func will break down the config override
+#   --config slurm.sbatch_args=[--time=05:00]
+# into key slurm.sbatch_args, value [--time=05:00]
+# And this func will return string in json form:
+# 'slurm': {'sbatch_args': ['--time=05:00']}
+def _dot_to_json(key: Any, value: Any) -> Any:
+    json_output: Any = {}
+    path = key.split(".")
+    target = reduce(lambda d, k: d.setdefault(k, {}), path[:-1], json_output)
+    target[path[-1]] = value
+    return json_output
+
+
+# Convert the NoneType value None to empty value {}
+# Some key may have value None and will raise TypeError if used.
+# For example:
+# TypeError: argument of type 'NoneType' is not iterable
+# TypeError: 'NoneType' object does not support item assignment
+#
+# Example config:
+# {'name': 'noop_single', 'checkpoint_storage': {...},
+#  ...
+#  'searcher': {...}, 'max_restarts': 0,
+#  'entrypoint': 'model_def:NoOpTrial',
+#  'slurm': None
+# }
+#
+# When called with key slurm.sbatch_args, the return config:
+# {'name': 'noop_single', 'checkpoint_storage': {...},
+#  ...
+#  'searcher': {...}, 'max_restarts': 0,
+#  'entrypoint': 'model_def:NoOpTrial',
+#  'slurm': {}
+# }
+def _fix_none_value_in_config(config: Dict[str, Any], key_path: List[str]) -> Dict[str, Any]:
     current = config
     for key in key_path[:-1]:
         current = current.setdefault(key, {})
-    current[key_path[-1]] = value
+
```

Comment:
does this work when the command line sets a value that is deeper than the first level?

e.g. with `hyperparameters:`, do `--config hyperparameters.optimizer_config.name=Adam` to set a nested hyperparameter.

if I am understanding the code right I suspect that'd throw a similar exception

Reply:
(determined) canmingcobble@Canmings-MacBook-Pro no_op % det e create --config slurm.sbatch_args="[--time=05:00]" --config hyperparameters.global_batch_size=32 --config hyperparameters.optimizer_config.name=Adam single.yaml .
Preparing files to send to master... 248.4KB and 29 files
Created experiment 772

The submitted configs

slurm:
  sbatch_args:
    - '--time=05:00'
hyperparameters:
  global_batch_size: 32
  optimizer_config:
    name: Adam

Reply:
So it works

Reply:
You are right, it does not work:

(determined) canmingcobble@Canmings-MacBook-Pro no_op % det e create --config slurm.sbatch_args="[--time=05:00]"  --config hyperparameters.optimizer_config.name=Adam single.yaml .
Traceback (most recent call last):
  File "/Users/canmingcobble/dev/determined-ee/harness/determined/cli/cli.py", line 288, in main
    parsed_args.func(parsed_args)
  File "/Users/canmingcobble/dev/determined-ee/harness/determined/cli/experiment.py", line 333, in create
    submit_experiment(args)
  File "/Users/canmingcobble/dev/determined-ee/harness/determined/common/api/authentication.py", line 475, in f
    return func(namespace)
  File "/Users/canmingcobble/dev/determined-ee/harness/determined/cli/experiment.py", line 239, in submit_experiment
    experiment_config = _parse_config_text_or_exit(config_text, args.config_file.name, args.config)
  File "/Users/canmingcobble/dev/determined-ee/harness/determined/cli/experiment.py", line 121, in _parse_config_text_or_exit
    config = parse_config_overrides(experiment_config, config_overrides)
  File "/Users/canmingcobble/dev/determined-ee/harness/determined/cli/command.py", line 349, in parse_config_overrides
    config = _fix_none_value_in_config(config, key.split("."))
  File "/Users/canmingcobble/dev/determined-ee/harness/determined/cli/command.py", line 309, in _fix_none_value_in_config
    current = current.setdefault(key, {})
AttributeError: 'NoneType' object has no attribute 'setdefault'
Failed to create experiment

Reply:
Fixed I hope.

det e create --config slurm.sbatch_args="[--time=05:00]"  --config hyperparameters.optimizer_config.name=Adam single.yaml .
Preparing files to send to master... 248.4KB and 29 files
Created experiment 783

```
slurm:
  sbatch_args:
    - '--time=05:00'
hyperparameters:
  optimizer_config:
    name: Adam
```